src/output.cpp

```diff
@@ -742,6 +789,149 @@ void TextOutput::map_hist(
   out_ << std::endl;
 }
 
+void TextOutput::map_tseries(
+    BPFtrace &bpftrace,
+    const BpfMap &map,
+    const TSeriesMap &values_by_key,
+    const std::vector<std::pair<KeyType, EpochType>> &latest_epoch_by_key) const
+{
+  map_tseries_contents(bpftrace, map, values_by_key, latest_epoch_by_key);
+  out_ << std::endl;
+}
+
+template <typename T>
+std::string tseries_to_str_text(BPFtrace &bpftrace,
+                                const TSeries &values,
+                                EpochType last_epoch,
+                                uint64_t interval_ns,
+                                uint64_t buckets)
+{
+  constexpr int graph_width = 53;
+  std::ostringstream res;
+
+  // We have no reference point from which to render the rest of the graph.
+  // Just print <no data> instead.
+  if (!last_epoch) {
+    res << "<no data>" << std::endl;
+    return res.str();
+  }
+
+  auto min_max = util::tseries_bounds<T>(values,
+                                         last_epoch - buckets + 1,
+                                         last_epoch);
+  T min_value = min_max.first;
+  T max_value = min_max.second;
+
+  // Ensure a range of at least three, so that we have concrete values for the
+  if (min_value == max_value) {
+    // Generally prefer if we can add some buffer on both sides of the points
+    // in the graph, but don't overflow if our min or max is at the upper
+    // bound of the range.
+    if (min_value > std::numeric_limits<T>::min()) {
+      min_value -= 1;
+    }
+    if (max_value < std::numeric_limits<T>::max()) {
+      max_value += 1;
+    }
+  }
+
+  std::string time_fmt = "%H:%M:%S";
+  std::string time_legend = "hh:mm:ss";
+  if (interval_ns <
+      static_cast<uint64_t>(std::chrono::nanoseconds(1us).count())) {
+    time_fmt += ".%k";
+    time_legend += ".ns       ";
+  } else if (interval_ns <
+             static_cast<uint64_t>(std::chrono::nanoseconds(1ms).count())) {
+    time_fmt += ".%f";
+    time_legend += ".us    ";
+  } else if (interval_ns <
+             static_cast<uint64_t>(std::chrono::nanoseconds(1s).count())) {
+    time_fmt += ".%l";
+    time_legend += ".ms ";
+  }
+
+  constexpr int padding = 21;
```

Comment:
Nit: the width and the padding (and any other random constants for tseries rendering) should maybe be part of some tseries config struct at the top of the function or in a header just for organizational sake - but non-blocking.

Reply:
IMO this feels like it should be part of a single refactor PR to move output constants into structs for the various types like lhist too, since there's some similar stuff going on in other places.

Reply:
No problem. My suggestion is non-blocking.